src/core/lib/iomgr/polling_interface.h

```diff
@@ -0,0 +1,167 @@
+// serves to abstract away that detail
+class PollableSet {
+ public:
+  // Add a Pollable to this collection
+  virtual void AddPollable(Pollable* pollable) = 0;
+
+ protected:
+  PollableSet() {}
+  ~PollableSet() {}
+
+ private:
+  friend class PollingJoin;
+  virtual void AddToPollingJoin(PollingJoin* join) = 0;
+};
+
+// A Poller is a set of file descriptors that a higher level item is
+// interested in, and a method to poll on them. For example:
+//    - a server will typically keep a poller containing all connected channels,
+//      so that it can find new calls to service
+//    - a completion queue might keep a poller with an entry for each transport
+//      that is servicing a call that it's tracking
+class Poller : public PollableSet {
+ public:
+  virtual ~Poller() {}
+
+  class Worker {};
+
+  // Do some work on a pollset.
+  // May involve invoking asynchronous callbacks, or actually polling file
+  // descriptors.
+  //
+  // Requires pollset's mutex locked. May unlock its mutex during its execution.
+  //
+  // worker is a (platform-specific) handle that can be used to wake up
+  // from grpc_pollset_work before any events are received and before the
+  // timeout has expired. It is both initialized and destroyed by
+  // grpc_pollset_work. Initialization of worker is guaranteed to occur BEFORE
+  // the pollset's mutex is released for the first time by grpc_pollset_work and
+  // it is guaranteed that it will not be released by grpc_pollset_work AFTER
+  // worker has been destroyed.
+  //
+  // It's legal for worker to be NULL: in that case, this specific thread can
+  // not be directly woken with a kick, but maybe be indirectly (with a kick
+  // against the pollset as a whole).
+  //
+  // Tries not to block past deadline. May call grpc_closure_list_run on
+  // grpc_closure_list, without holding the pollset lock
+  virtual grpc_error* Poll(grpc_exec_ctx* exec_ctx, Worker** worker,
+                           grpc_millis deadline) GRPC_MUST_USE_RESULT = 0;
+
+  // Break one polling thread out of polling work for this pollset.
+  // If specific_worker is non-NULL, then kick that worker.
+  virtual grpc_error* Kick(grpc_exec_ctx* exec_ctx,
+                           Worker* specific_worker) GRPC_MUST_USE_RESULT = 0;
+
+  // Begin shutting down the pollset, and call on_done when done.
+  // pollset's mutex must be held
+  virtual void Shutdown(grpc_exec_ctx* exec_ctx, grpc_closure* on_done) = 0;
+
+  // CHANGING: return the size of a Poller with the current polling engine
+  // (will be moved to the polling engine interface directly in a future
+  // revision)
+  static size_t PollerSize();
+
+  // CHANGING: instantiate a Poller with the current polling engine
+  static Poller* Create(void* memory, gpr_mu** mu);
+
+ protected:
+  Poller() {}
+
+ private:
+  void AddToPollingJoin(PollingJoin* polling_join) override final;
+};
+
+// A PollingJoin serves to join multiple pollers and pollables.
```

Comment:
It would probably be a good idea to document the fact that this join only ever adds to the Pollers (or PollableCollections), it never reads from them.  In other words, if you add a PollableCollection to a PollingJoin, any Pollables contained in the PollableCollection are *not* added to any other PollableCollections (Pollers or PollingJoins) that were previously added to the PollingJoin.  Only Pollables added directly to the PollingJoin are added to the PollableCollections within the PollingJoin.